File: python/vcp_sdk/identity.py
"""Capability identity and argument binding (SPEC §4, §7, §8).

Capability identity is the hash of the capability *contract* — the
security-relevant subset of the manifest. Per §4 the contract MUST include
``issuer``, ``name``, ``version``, ``input_schema``, ``output_schema``,
``effects``, ``determinism``, and ``sandbox``. It MUST NOT include the display
summaries, signatures, or provenance.

Because JCS sorts object keys, the *order* the contract fields are listed in
here does not affect the hash; only the *set* of fields (the partition) and
their values matter. This module reproduces ``conformance/vectors/
capability-identity.json`` and ``argument-binding.json`` exactly.
"""

from __future__ import annotations

from typing import Any, Mapping

from .canonical import hash as _hash

__all__ = [
    "CONTRACT_FIELDS",
    "extract_contract",
    "contract_hash",
    "capability_id",
    "argument_hash",
    "parse_capability_id",
]

# The normative contract partition (SPEC §4 / manifest.schema.json).
CONTRACT_FIELDS = (
    "issuer",
    "name",
    "version",
    "input_schema",
    "output_schema",
    "effects",
    "determinism",
    "sandbox",
)


def extract_contract(manifest_or_capability: Mapping[str, Any]) -> dict:
    """Pull the security-relevant contract subset out of a manifest.

    Accepts either a full manifest (``{"issuer": ..., "capability": {...}}``)
    or a flat mapping that already carries the contract fields (such as the
    ``contract`` object in the conformance vector). ``issuer`` may live at the
    top level of a manifest while the rest live under ``capability``.
    """
    m = manifest_or_capability
    cap = m.get("capability", m)
    contract: dict[str, Any] = {}
    for field in CONTRACT_FIELDS:
        if field in m and field not in cap:
            contract[field] = m[field]
        elif field in cap:
            contract[field] = cap[field]
        else:
            raise KeyError(f"contract field missing: {field!r}")
    return contract


def contract_hash(manifest_or_contract: Mapping[str, Any]) -> str:
    """``sha256(JCS(contract))`` as ``sha256:<hex>`` (SPEC §4).

    If the mapping already looks like a bare contract (carries every contract
    field and no ``capability`` envelope) it is hashed directly; otherwise the
    contract subset is extracted first.
    """
    m = manifest_or_contract
    if "capability" not in m and all(f in m for f in CONTRACT_FIELDS):
        contract = {f: m[f] for f in CONTRACT_FIELDS}
    else:
        contract = extract_contract(m)
    return _hash(contract)


def capability_id(manifest_or_contract: Mapping[str, Any]) -> str:
    """``vcp:cap:<name>@<contract_hash>`` (SPEC §4)."""
    ch = contract_hash(manifest_or_contract)
    m = manifest_or_contract
    name = m.get("name") or m.get("capability", {}).get("name")
    if not name:
        raise KeyError("contract is missing 'name'")
    return f"vcp:cap:{name}@{ch}"


def argument_hash(arguments: Mapping[str, Any]) -> str:
    """``sha256(JCS(arguments))`` as ``sha256:<hex>`` (SPEC §7, §8)."""
    return _hash(arguments)


def parse_capability_id(cap_id: str) -> tuple[str, str]:
    """Split ``vcp:cap:<name>@sha256:<hex>`` into ``(name, "sha256:<hex>")``."""
    if not cap_id.startswith("vcp:cap:"):
        raise ValueError(f"not a capability id: {cap_id!r}")
    rest = cap_id[len("vcp:cap:") :]
    name, _, digest = rest.partition("@")
    if not digest:
        raise ValueError(f"capability id missing @hash: {cap_id!r}")
    return name, digest

File: rust/crates/vcp-sdk/src/identity.rs
//! Content-addressed capability identity (VCP §4) and argument binding (§7/§8).
//!
//! `contract_hash = sha256(JCS(contract))`
//! `capability_id = "vcp:cap:" + name + "@" + contract_hash`
//! `argument_hash = sha256(JCS(arguments))`
//!
//! The `contract` is the security-relevant subset of a manifest: `issuer`,
//! `name`, `version`, `input_schema`, `output_schema`, `effects`, `determinism`,
//! `sandbox`. Display strings, signatures, and provenance are excluded so they
//! can change without altering identity. Any change to a contract field yields a
//! new identity (rug-pull becomes a visible diff).

use serde::Serialize;
use serde_json::Value;

use crate::jcs;

/// Compute `contract_hash` = `sha256(JCS(contract))` for an arbitrary contract
/// value (the security-relevant subset of a manifest, already partitioned).
pub fn contract_hash_value(contract: &Value) -> String {
    jcs::hash_value(contract)
}

/// Compute `contract_hash` from any `Serialize` contract type.
pub fn contract_hash<T: Serialize>(contract: &T) -> Result<String, serde_json::Error> {
    jcs::hash(contract)
}

/// Build the content-addressed `capability_id`:
/// `vcp:cap:<name>@<contract_hash>`.
pub fn capability_id(name: &str, contract_hash: &str) -> String {
    format!("vcp:cap:{name}@{contract_hash}")
}

/// Compute both `contract_hash` and `capability_id` from a contract value and
/// its `name`.
pub fn identity_for(name: &str, contract: &Value) -> (String, String) {
    let ch = contract_hash_value(contract);
    let id = capability_id(name, &ch);
    (ch, id)
}

/// `argument_hash = sha256(JCS(arguments))` (§7, §8). A grant binds to this; if
/// the Planner changes any argument the hash no longer matches.
pub fn argument_hash_value(arguments: &Value) -> String {
    jcs::hash_value(arguments)
}

/// Compute `argument_hash` from any `Serialize` arguments type.
pub fn argument_hash<T: Serialize>(arguments: &T) -> Result<String, serde_json::Error> {
    jcs::hash(arguments)
}

/// Extract the `sha256:<hex>` digest embedded in a `vcp:cap:<name>@sha256:<hex>`
/// identifier, if present. Comparison of identifiers elsewhere is exact, byte
/// for byte (§3) — this helper is only for confirming `id` carries the recomputed
/// `contract_hash`.
pub fn digest_of(capability_id: &str) -> Option<&str> {
    capability_id.split_once('@').map(|(_, hash)| hash)
}
